### ISSUE TITLE
Capture FAST headings instead of LCSH headings in MARC input stream

### ISSUE THREAD

**Issue by mcritchlow** [ucsdlib/damsmanager#104]
Please ping @ucsdlib/domm for any questions.

If a MARC record contains a 6XX field:

Whose 2nd indicator = 7 AND contains the subfield:'|2fast'

Then load that 6xx type field ONLY and do not load a matching 6xx field with a second indicator of 0.

Ex:

650  0 Architecture, Modern|y20th century|vGuidebooks   DO NOT LOAD
650  0 Architect-designed houses|zCalifornia|vGuidebooks  DO NOT LOAD
**650  7 Architect-designed houses.|2fast|0(OCoLC)fst00813106 LOAD**
**650  7 Architecture, Modern.|2fast|0(OCoLC)fst00813842 LOAD**
**650  7 Themes, motives.|2fast|0(OCoLC)fst01355139 LOAD**
651  0 California|vGuidebooks   DO NOT LOAD
**651  7 California.|2fast|0(OCoLC)fst01204928   LOAD**
**655  7 Guidebooks.|2fast|0(OCoLC)fst01423871 LOAD**

Instruction: Load the **bold** fields above.  Don't load the others.

For ingest:
- Terms should be matched against the existing database using 1) the fast ID - subfield |0 when available, if not available 2) using the text/string value
- When the term already exists in the DAMS, that ark/term should be referenced
- When the term does not already exist in the DAMS, a record for the new term should be created, including encoding the fast ID |0

**Comment by lsitu** [comment]
@ucsdlib/domm Could I have an example that can be used for testing? And the corresponding converted result if possible. Thanks.

**Comment by arwenhutt** [comment]
@pitviper6 Do you have time to help @lsitu with a good example record he can test with?

**Comment by GregReser** [comment]
I don;t have a converted version, but this Roger record cold be a good example.  It has both FAST and non-FAST subjects with the same labels.
http://roger.ucsd.edu/record=b8529446~S9

**Comment by arwenhutt** [comment]
@lsitu the record Greg links to has a few different subject types included (648, 650, and 655), do you need something which covers all subject types, or is it sufficient to have a few for testing and then use that pattern in the stylesheet's other mappings?  

For a record with all types, I think we'll have to get an artificial test record created by our MARC catalogers.

**Comment by lsitu** [comment]
@arwenhutt Yes, It's good to list all subject types that we want to apply the fast conversion rule. And I am also wondering how these subject type are converted to MARCXML at these time so that we can decide how to update JollyRoger to convert these fields to MARCXML.

**Comment by lsitu** [comment]
@arwenhutt @GregReser From the example http://lib-hydratail-staging.ucsd.edu:8080/jollyroger/get?type=bib&mods=true&ns=true&value=b8529446 MarcXML, I don't see any fast id included so I think we may need to update JollyRoger to include all field we want as the first step. Do you have any thoughts on the MarcXML that is used for dams4 RDF conversion?

**Comment by arwenhutt** [comment]
The DAMS field that each type of subject (648, 650, 655, etc.) maps to should not be changed as part of this ticket.  What we want to change is to ingest the values from the ones that contain FAST values (marked by a second indicator 7 AND contains the subfield:'|2fast') instead of the ones that contain LCSH values (marked by a second indicator 0).  Currently we import only the LCSH ones and not the FAST.  We want to switch this.

**Comment by arwenhutt** [comment]
@lsitu I don't know what parts of the current MARC conversion are happening at the MARC->MARCXML part vs. the MARCXML->DAMS 4 conversion, @pitviper6 may know some, but she came on after this was already in the works, so I'm afraid the only authoritative source would be Esme. 

My memory is that the MARC->MARCXML was based on a LOC stylesheet but with local customization. 

It seems that the safest way forward, for now, is to update the MARC->MARXML transformation to include the FAST ids

**Comment by lsitu** [comment]
@arwenhutt Yes. It sounds good to me. But what format do you suggest for the id field in the MARCXML and how it looks like in DAMS4 RDF/XML?

**Comment by arwenhutt** [comment]
@lsitu ah, so your question is where do we map the ID value?  

From the perspective of the MADS data model, I think it maps to mads:hasExactExternalAuthority, e.g. <mads:hasExactExternalAuthority rdf:resource="http://id.loc.gov/authorities/names/no2008083372"/>

But we haven't done this much in D4 so I'm not sure how completely this is modeled out in D4

I found the name example above but looking for more is a needle in a haystack.

Is mads:hasExactExternalAuthority already available in D4 for subjects?

**Comment by lsitu** [comment]
@arwenhutt Yes. I think we've got it in DAMS4 for authority records, including subjects.

**Comment by arwenhutt** [comment]
Great!  then that's the place.

**Comment by lsitu** [comment]
@arwenhutt For the following 650 field in the example above:
<datafield tag="650" ind1=" " ind2="7">
<subfield code="a">Architect-designed houses.</subfield>
<subfield code="2">fast</subfield>
<subfield code="0">(OCoLC)fst00813106</subfield>
</datafield>

What's the MARCXML looks like? Do we need to convert the fast id (OCoLC)fst00813106 to a URI or not since mads:hasExactExternalAuthority should link to an external authority?

**Comment by arwenhutt** [comment]
For the ID encoding, yes, it would be good to convert to a URI.  The URI for the example you gave is `http://id.worldcat.org/fast/813106` 

It looks like they aren't using the leading zero's in the URI (though if you include them it still resolves to the same record, but including them could cause problems in future matching)

For the MARCXML, I don't know exactly how/when the MARCXML is generated locally (the link to the sample you sent above is actually MODS output not MARCXML), but it should be something like this:
```
<datafield tag="650" ind1=" " ind2="7">
<subfield code="a">Architect-designed houses.</subfield>
<subfield code="2">fast</subfield>
<subfield code="0">(OCoLC)fst00813106</subfield>
</datafield>
```

**Comment by lsitu** [comment]
Thanks @arwenhutt. Yes, I mean the MODS output looks like, which is used for DAMS4 RDF conversion.

**Comment by arwenhutt** [comment]
According to the sample you sent the MODS output currently being created conforms to the MODS 3.5 standard, and according to that the fast ID value should (I believe) be encoded as @valueURI 

The current MODS output doesn't appear to include the IDs, so presumably it would have to be added to the MARCXML->MODS transformation.

**Comment by lsitu** [comment]
@arwenhutt Do we need to support the mads:hasExactExternalAuthority field with the CSV format in the Excel Input Stream?

**Comment by arwenhutt** [comment]
@lsitu Nope, for the excel input stream we are just matching against existing subject records in the DB.  Also the subject import supports it mads:hasExactExternalAuthority field so we can add subjects with URIs via that.  

We will likely want to rethink all of this when we redesign and rebuild the import tools in D5, but it is not necessary now in the D4 Excel Input Stream.

**Comment by lsitu** [comment]
@arwenhutt Okay. I'll move ahead to update JollyRoger and the MARC/MODS Import tool to support it them.

**Comment by lsitu** [comment]
@arwenhutt @GregReser Attached is the sample record for b8529446. You can simply rename it to .xml to review it. Thanks.
[export-b8529446.txt](https://github.com/ucsdlib/damsmanager/files/537097/export-b8529446.txt)

**Comment by lsitu** [comment]
@mcritchlow I've created PR https://github.com/ucsdlib/damsmanager/pull/106 to add FAST headings support to the MARC/MODS Input Stream, which will extract the FAST headings if it exists otherwise the LCSH will be extracted. I think it's ready for review now. Thanks.

**Comment by mcritchlow** [comment]
@arwenhutt @GregReser - how sure are we that the subfield value for FAST will always be `fast` in lowercase. Are we sure it will never be `FAST` or `Fast`? I'm wondering if we should, or need to, force lowercase on our end.

**Comment by arwenhutt** [comment]
@mcritchlow good question, I'll check in with our some of our marc experts tomorrow and see.  But if we could easily accommodate any casing, that would probably be safer. 

Also, @lsitu I'll look at the test record tomorrow.

**Comment by GregReser** [comment]
Agreed about forcing "FAST" to lower case - just in case.
The sample looks good to me.
Are we also bringing in non-fast 6xx field with a second indicator of 0 IF there is not a matching fast 6xx?

Example 6xx from Roger bib# b3437818:

650  0 Interior decoration|xHistory 
650  0 Interior architecture|xHistory 
650  0 Interior decoration in art 
653 0  Interiors|aDecoration|aHistory 
655  7 History.|2fast|0(OCoLC)fst01411628 

If we only load the 655 7 "History" above, we'd be missing out on a lot of useful subjects.
The logic could be:
If a MARC record contains a 6XX field:

Whose 2nd indicator = 7 AND contains the subfield:'|2fast' ('fast' = toLowerCase)

Then load that 6xx type field ONLY and do not load a 6xx field **with matching first subfield** AND with a second indicator of 0.

We cant match on |a because the 653 uses multiple |a

**Comment by lsitu** [comment]
@mcritchlow I've updated the PR to support case intensive FAST in the style sheet.

**Comment by lsitu** [comment]
@GregReser It's hard to support this rule since we convert the MODS output to DAMS4 RDF at this time. In the MODS format output, all those 6xx fields are converted to mods:subject and they are mixed with the fields that are marked as "fast". So I think we may need to convert all subject fields to FAST with an ID, or at lease marked them as "fast".

**Comment by GregReser** [comment]
@lsitu do you mean that the input stream would search FAST for a match and use it?  In other words it would reconcile the non-FAST subjects at the time of ingest?  That would be great.

**Comment by lsitu** [comment]
@GregReser Yes. So we would support both FAST subjects and non-FAST subject MARC records at the same time, and no duplicates as well.

**Comment by GregReser** [comment]
@lsitu can you test this on b3437818?

**Comment by lsitu** [comment]
@GregReser Here it is:
[b3437818-export.txt](https://github.com/ucsdlib/damsmanager/files/539535/b3437818-export.txt)

**Comment by arwenhutt** [comment]
@lsitu For now please do not include non-FAST headings in the MARC input.  Thanks.

**Comment by mcritchlow** [comment]
Given @arwenhutt most recent comment, i think @lsitu's PR is ready to go. I'll review and merge now. Thanks all!

**Comment by arwenhutt** [comment]
Not sure if it's relevant, but I'm QA'ing the output now, it looks good so far.

**Comment by mcritchlow** [comment]
@arwenhutt - we can certainly wait for that, sorry i missed that comment :)

**Comment by arwenhutt** [comment]
No problem! Okay, I'm done :) @lsitu the output in this comment https://github.com/ucsdlib/damsmanager/issues/104#issuecomment-254601656 looks good. 

Since the sample does only include some of the 6xx fields, I only verified that those were coming in as expected. 

We can do a more thorough test, using a validation record like we discussed previously, but there isn't any variation in handling of the different terms, so I'm not sure it's necessary.

@lsitu :
- Can you verify that the same mapping logic was applied to all the 6xx/subject fields in the transformation?  
- Is there any variation in the code that you think would make it useful to test a complete validation record?

**Comment by arwenhutt** [comment]
Also, I forgot to add, we'll need to test the matching against the DAMS values (bullets 1&2 in the original ticket) functionality after it's deployed to staging.

**Comment by lsitu** [comment]
@arwenhutt For your questions above:
> Can you verify that the same mapping logic was applied to all the 6xx/subject fields in the transformation?

I think the logic should be applied to all 6xx/subject fields. But do we have a FAST id for hierarchicalGeographic or not? Also, do we have other fields that are not 6xx but map to subject as well?

>Is there any variation in the code that you think would make it useful to test a complete validation record?

No. The codes are general for all subjects.

**Comment by arwenhutt** [comment]
> I think the logic should be applied to all 6xx/subject fields. 
Great, sounds good!

> But do we have a FAST id for hierarchicalGeographic or not? 
Could you clarify what you mean?  This could be interpreted a number of ways, so I'm not sure which question to answer.  

> Also, do we have other fields that are not 6xx but map to subject as well?  
We don't have documentation of the mapping from MARC to DAMS, just the scripts themselves (which were customized from community scripts).  So I couldn't say authoritatively one way or the other.  BUT it seems unlikely (too potentially messy I'd think), and even more unlikely that if there other fields mapped, that they would contain either LCSH precoordinated headings (which we don't want).  Probably the best way to check this would be looking at the scripts themselves.

I think the initial mapping looks good though, so probably ready for @mcritchlow to review and merge.

**Comment by lsitu** [comment]
@arwenhutt Unlike other subject fields, hierarchicalGeographic subject is handle in a separate block in MODS output conversion with no id converted. This is the only variation I see and I may need to add id conversion logic for it if it do have an id.
For fields other than 6xx that are converted to subjects, I think we may need to marked them as FAST headings as well.

**Comment by arwenhutt** [comment]
> Unlike other subject fields, hierarchicalGeographic subject is handled in a separate block in MODS output conversion with no id converted. 


@lsitu I don't think we need to worry about the hierarchicalGeographics headings.  I checked with Ryan Finnerty (head of the database maintenance unit for our ILL) about how common they (662 fields specifically) are in MARC records and his exact quote was "I don't believe I have ever encountered one of those in my life".  So that's good. 

> For fields other than 6xx that are converted to subjects, I think we may need to marked them as FAST headings as well. 


We don't want to mark items as FAST headings unless they are FAST headings.  Can you clarify what non 6xx fields are converted to subjects, so I can investigate?

**Comment by lsitu** [comment]
@arwenhutt I think only the following non 6xx fields could be converted to subjects by JollyRoger:
marc:datafield[@tag=034] createSubGeoFrom034 => cartographics
marc:datafield[@tag=045] createSubTemFrom045 => temporal
marc:datafield[@tag=255] createSubGeoFrom255 => cartographics
marc:datafield[@tag=752] createSubGeoFrom662752 => hierarchicalGeographic

**Comment by arwenhutt** [comment]
Okay, I don't think those will have FAST headings assigned by OCLC, so they won't be available in the MARC.  So I think for initial implementation we can leave them.  

We'll do an evaluation of the process with some of our MARC folks and based on that look at whether we need to make any modifications to the mapping or processing for D5.

**Comment by arwenhutt** [comment]
@GregReser I verified that this has been deployed to staging and is ready for testing.  Please test that it works as expected, particularly the new functionality of matching against existing dams records using the FAST ID, and if that is not available, the text string (processing logic outlined above).  
Thanks!

**Comment by GregReser** [comment]
@lsitu Testing in staging - 
Roger record = b3582973
DAMS object = https://librarytest.ucsd.edu/dc/object/bd76949215

The Roger record has 3 non-FAST subjects.  I believe three new DAMS ARKs were created for these on ingest.  These should not have imported, right?

650  0 Architecture, Domestic|xHistory 
650  0 Dwellings 
650  0 Historic buildings 

http://library.ucsd.edu/ark:/20775/bd3804101n
http://library.ucsd.edu/ark:/20775/bd51010413
http://library.ucsd.edu/ark:/20775/bd3804101n

------------------------------------------------
-------------------------------- Thu Oct 27 13:04:41 PDT 2016 --------------------------------
Object Import status:
[Title]   -   [URI]   -   [Status]   -   [Timestamp]
Created derivatives for http://library.ucsd.edu/ark:/20775/bd76949215/1.PNG (https://lib-hydratail-staging.ucsd.edu:8443/dams/api/files/bd76949215/1.PNG/derivatives?ts=dams&priority=4&user=greser&client=MARC+staging-0-g148e935c7587bf147823a27ee09cec0ec5d02e24).
Created zoomify tiles http://library.ucsd.edu/ark:/20775/bd76949215/1.PNG (https://lib-hydratail-staging.ucsd.edu:8443/dams/api/files/bd76949215/1.PNG/derivatives?ts=dams&priority=4&user=greser&client=MARC+staging-0-g148e935c7587bf147823a27ee09cec0ec5d02e24).

The book of houses - http://librarytest.ucsd.edu/dc/object/bd76949215 - successful - 2016-10-27T13:04:50

______________________________________________________________________________________________
Import summary: successful - Imported 1 objects in 1 metadata file 
- Total 6 records ingested. 
- Total 1 files ingested.

The following 1 object is found in the source metadata: 
http://librarytest.ucsd.edu/dc/object/bd76949215 	b3582973.rdf.xml

Found 10 records. The following 6 records are imported: 
http://library.ucsd.edu/ark:/20775/bd1244351s
http://library.ucsd.edu/ark:/20775/bd25412891
http://library.ucsd.edu/ark:/20775/bd3804101n
http://library.ucsd.edu/ark:/20775/bd51010413
http://library.ucsd.edu/ark:/20775/bd6397981z
http://library.ucsd.edu/ark:/20775/bd76949215

**Comment by lsitu** [comment]
@GregReser I don't see the marc Record [b3582973](http://lib-hydratail-prod.ucsd.edu:8080/jollyroger/get?type=bib&mods=false&ns=true&value=b3582973) that you used for testing includes any FAST headings, so it will be treated as-is with all subjects ingested. To exclude those non-FAST 6xx fields from ingested, we have to set at least one 6xx field to use FAST headings.

**Comment by GregReser** [comment]
@arwenhutt Need to confirm the meaning of your earlier comment "For now please do not include non-FAST headings in the MARC ingest" 

1. Non-FAST headings should not be imported, regardless of if they have a matching FAST heading.  If a MARC record contains only non-FAST headings, then no headings will be imported.

OR 

2. Non-FAST headings with no matching FAST heading should be treated as they currently are (string match to DAMS ARK > ELSE create new ARK)

**Comment by arwenhutt** [comment]
This: 

> Non-FAST headings should not be imported, regardless of if they have a matching FAST heading. If a MARC record contains only non-FAST headings, then no headings will be imported. 

Specifically drawing from the text in the original issue:

> Don't load the others.

The primary concern of this ticket is to update the ingest so that no LCSH precoordinated headings are ingested, so that once the subject conversion is complete, we will not be reintroducing complex subjects (in content or in encoding).   There are always things that may fall in grey areas (what if they are MESH headings?  or other non LCSH headings?  etc.) but this was the approach discussed at the 6/21 "FAST headings from MARC records - DAMS ingest discussion" meeting (Matt P., Ryan F., Michelle Mascaro, Juliane Schneider, Ryan J., Greg, and myself) and reviewed in JIRA https://lib-jira.ucsd.edu:8443/browse/DM-197 .  

Once we have the tool updated per this scope we'll revisit with the MARC experts to 1) make sure that we're on track with the non-DAMS dependencies 2) evaluate the initial tool 3) identify any modifications that are needed.  Any modifications that are identified would then be filed as new issue requests.  E.g. the suggestion to string match to existing DAMS ark (minus the > ELSE create new ARK) could be a very good and "safe" addition post subject conversion.  For now though it would just propagate LCSH headings.

**Comment by GregReser** [comment]
@lsitu FAST personalName and corporateName are not being imported.  I test twice to confirm.
Roger record = b3052904
DAMS object = http://librarytest.ucsd.edu/dc/object/bd3087371f AND http://librarytest.ucsd.edu/dc/object/bd9845111x

The MARC record has FAST terms:
610 27 E.W. Scripps (Ship)|2fast|0(OCoLC)fst00774418 
610 27 Scripps Institution of Oceanography.|2fast|0(OCoLC)fst00540219

The DAMS objects have empty properties:

<dams:relationship>
  <dams:Relationship>
	<dams:role rdf:resource="http://library.ucsd.edu/ark:/20775/bb67931665"/>
	<dams:personalName/>
  </dams:Relationship>
</dams:relationship>

<dams:personalName/>
<dams:corporateName/>

**Comment by lsitu** [comment]
@GregReser I've updated it to capture only those FAST subject and there is no non-FAST subjects extract when there are no non-FAST headings exist. So we only expect to ingest those MARC records that are updated to have FAST headings.
I also fix the name issue you found with the change. Could you try it again? Thanks.

**Comment by GregReser** [comment]
@lsitu I'm out of the office today, but I will test it first thing Monday.

**Comment by lsitu** [comment]
@GregReser Thanks. Don't worry about it. I think we've already been in advance for the sprint deployment.

**Comment by GregReser** [comment]
@lsitu Confirmed it is functioning as expected - for the following MARC headings, only FAST headings were imported and FAST ID was included in the DAMS authority.
600 > PersonalName
610 > CorporateName
650 > Topic
651 > Geographic
655 > GenreForm

I only found one problem:
GenreForm value should not include dot at end.  See how topic and geographic values are handled.
<mads:GenreFormElement>
	<mads:elementValue>Filmed speeches.</mads:elementValue>
<mads:TopicElement>
	<mads:elementValue>Presidents</mads:elementValue>	
<mads:GeographicElement>
	<mads:elementValue>United States</mads:elementValue>

Test records:
http://roger.ucsd.edu/record=b3052904 > https://librarytest.ucsd.edu/dc/object/bd61590712 
http://roger.ucsd.edu/record=b1772000 > https://librarytest.ucsd.edu/dc/object/bd7729054w
http://roger.ucsd.edu/record=b3582973 > https://librarytest.ucsd.edu/dc/object/bd18928248
http://roger.ucsd.edu/record=b4090110 > https://librarytest.ucsd.edu/dc/object/bd1210224n
http://roger.ucsd.edu/record=b9296620 > https://librarytest.ucsd.edu/dc/object/bd2780204c

**Comment by lsitu** [comment]
@GregReser I've update to remove the leading and ending punctuations from subject GenreForm. I also fixed the missing FAST ID and scheme issues with GenreForm. Could you take a look again? Thanks.

**Comment by GregReser** [comment]
@lsitu Looks good.
Testing if an existing DAMS authority is matched on ingest:
1. Existing DAMS authority WITH FAST ID
 - yes, the ingested term "Filmed speeches" was matched to existing DAMS authoritybd2643682c

2. Existing DAMS authority WITHOUT FAST ID
 - no, a new DAMS authority was created.  This results in duplicate authorities, one with FAST ID and one without FAST ID.  I think the MARC FAST heading should be string matched to the existing DAMS authority and the FAST ID should be added.

"Marine biology"
Roger MARC ingest =  http://roger.ucsd.edu/record=b1767148
Existing DAMS authority for "Marine biology"  = http://library.ucsd.edu/ark:/20775/bb2424682s
New ingested DAMS object = http://librarytest.ucsd.edu/dc/object/bd8002093j
New DAMS authority for "Marine biology" = http://library.ucsd.edu/ark:/20775/bd5442345g

**Comment by lsitu** [comment]
@GregReser Will the updating of existing authority records to FAST heading is another bigger issue that is not related to the FAST heading support in this ticket?
I think we are ensured that the authority records are matching by its label and the mads scheme like FAST, lcsh, lcgft etc. But let me know if you would like to change this mapping rule for authority records lookup and linking so that we can create a ticket for it. Thanks.

**Comment by GregReser** [comment]
@lsitu Oh right, the FAST ID does not need to be added to an existing DAMS authority for this ticket, so disregard that.  I think the instructions do say that if a MARC FAST term is string matched to an existing DAMS ARK/term, that ARK should be used.  So in my above test, a new ARK should not have been created.  The existing ARK should have been used.  Am I right about that?

From the instructions:
Terms should be matched against the existing database using 1) the fast ID - subfield |0 when available, if not available 2) using the text/string value
When the term already exists in the DAMS, that ark/term should be referenced

**Comment by lsitu** [comment]
@GregReser I think we may need to talk about how an authority records should be looked up and link to an object a little more before implement a matching rule. Here is what we are doing now:
1. Matching the model (a class, like mads:Topic, mads:PersonalName, mads:GenreForm etc.)
2. Matching the mads scheme (such as lcsh,  and fast in this case)
3. Matching the authoritative label
4. Match other properties (such as the fast ID in this case)

Since we don't want to match #2 here, does the following matching rules sounds good to you?
1. Matching the model 
2. Matching the authoritative label
3. Match other properties, in this the FAST ID if it exists
For authority records that don't have a FAST ID, should we only look up and match it to an authority record that doesn't have a FAST ID only, or simply match it to any authority records that have the same authoritative label that may have a fast id and a mads scheme?

**Comment by GregReser** [comment]
@lsitu For matching, the instructions on this ticket don't mention matching the model.  Do you mean that when no FAST ID exists and a string match is made, the DAMS mads class must match the MARC tag type?  For example, 
IF MARC 650 == mads:Topic, match that ARK
ELSE create a new GenreForm ARK?

**Comment by lsitu** [comment]
@GregReser No. I think we may need to talk about a more general way to lookup an authority record for matching. It seems like that the model (class) is the key to compare two concepts. For example, a PersonalName and a Topic with the same literal value should be considered different concepts. So I believe we should match the model in the first place, then we can consider match other properties we want, such as the authoritative label, the FAST ID etc. While matching an authority record with no FAST ID is a special case of this that we should discuss.
What do you think?

**Comment by GregReser** [comment]
@lsitu The problem is that that we sometimes use a different mads class than the FAST type.  It is very possible that a FAST 655 will match to a topic not genreForm.  We have seen this in the subject conversion project.  We are changing some of these mismatches when they are completely wrong (personalName != geographic) but we are choosing to use a different class in some cases (event == topic).

I think, for now, we should not match on class.  It would be better to make a match to an existing ARK since we will be fixing those ARKS in the subject conversion project (that process has not been worked out yet, so I'm not sure about this).  Also, if we decided to use a FAST genre as a topic, we would want to match that and would not want a new genreForm authority for the same concept to be made.

@arwenhutt what do you think?

**Comment by arwenhutt** [comment]
@GregReser 
I think this sounds like a good approach.  This is complicated, and I don't think there's a 100% solution, but I agree that we're going to get much better results by prioritizing FAST ID and then string label matches.  I could see utilizing the class _if_ the ID or string/label matched more than one existing record in the DAMS.  But I believe (once the subjects and names have been cleaned up) that will be unusual, so I'm not sure it's worth planning for. 

1. Match the FAST ID if it exists
2. Match the authoritative label
3. If, multiple DAMS records are returned, match the class

**Comment by lsitu** [comment]
@arwenhutt @GregReser I think we should have some duplicate authority records in dams4 that need to clean up later since we haven't cleaned them up yet. Will it be better to just separate the matching rule implementation and the clean up work that we need to do?

**Comment by GregReser** [comment]
@lsitu Yes, I think we should do the clean up later.  We will need to do some further analysis after the first round of subject conversion to see what further we needs to be done.

**Comment by arwenhutt** [comment]
Agreed.  There shouldn't be clean up as part of the work on this ticket.

**Comment by lsitu** [comment]
@arwenhutt @GregReser Once the new matching rules implementation in DAMS Manager, we'll get the following results for records linking:
1. The new matching rules is a global change for authoritative records lookup and linking, which will apply to both the Excel Input Stream and the MARC/MODS Input Stream.
2. By matching the mads:authoritativeLabel without matching the model (class), it could bring up any records in the database that has the mads:authoritativeLabel with the same literal value. These include those records created by mistake in the past few years that has the same mads:authoritativeLabel since we never cleaned it up. And those dams subjects that has the mads:authoritativeLabel like scientificName, Series etc will be hit for matching as well.

If this is not the result you expected, we may need to adjust the matching rules a little bit. Otherwise I would just move ahead for implementation. Thanks.

**Comment by arwenhutt** [comment]
@lsitu thanks for that information.  If this is a global rule then I think we'll need to discuss this a little more. 

@GregReser I think we should probably discuss this at our unit meeting tomorrow.  If you have time to prepare, I'd like you to lead that conversation on what the implications of making a global change like this are and what the alternatives are.  If not, let me know and I'll look at rescheduling our meeting to accommodate.  Thanks!

**Comment by GregReser** [comment]
@lsitu DOMM discussed the options and came up with a proposed global matching logic.  I had to write this a little differently to help me keep the steps straight.  Let me know if you see problems or if it is unclear.

1. IF the source term has an ID AND it matches the ID in a single DAMS authority THEN link to that ARK
2. ELSE IF the source term label AND type match a DAMS authority authoritative label AND class THEN link to that ARK
3. ELSE create a new DAMS authority WHERE DAMS class = source term type, authoritative label = label

Question: after this change to the MARC input stream, will the “Preview converted source in CSV format” file contain only FAST headings?  I would think so, since the CSV is created from the converted RDF/XML.

**Comment by lsitu** [comment]
@GregReser That looks good and I am assuming that the ID here is the FAST heading ID.
For your question, I'll say Yes but you won't see any FAST heading ID in the CSV export file.

**Comment by GregReser** [comment]
@lsitu For the global matching, it would be any ID.  We are looking at using LoC, VIAF, and other IDs in the subject conversion and will most likely allow these for new ingests in the future (with priority rules).
For the MARC input stream, the logic, for now at least, should be to only import 2nd indicator = 7 AND contains the subfield:'|2fast'.

**Comment by lsitu** [comment]
Hi Greg,
Thank you very much for the clarification. So if I use the value in properties mads:hasExactExternalAuthority as the ID for authority look up, all IDs will be included. Is it correct?
For the MARC input stream, I won't do anything for FAST metadata extraction this round but only change the rules for authority records lookup, so you will see the same metadata. The only difference is that whether the authority records are created new or using the existing ones.

**Comment by GregReser** [comment]
Correct, all IDs in mads:hasExactExternalAuthority will be included.  We anticipate doing periodic analysis of the DAMS subjects to correct duplicate and incorrect authorities.

So, to see where we are with finishing this ticket...
In the last test I did, a new DAMS record was created for the MARC 650 FAST heading "Marine biology".  I think that should not have happened because the existing DAMS record with label "Marine biology" is a topic.  The label and the class match, so it should have linked to the existing ARK.

"Marine biology"
Roger MARC ingest = http://roger.ucsd.edu/record=b1767148
Existing DAMS authority for "Marine biology" = http://library.ucsd.edu/ark:/20775/bb2424682s
New ingested DAMS object = http://librarytest.ucsd.edu/dc/object/bd8002093j
New DAMS authority for "Marine biology" = http://library.ucsd.edu/ark:/20775/bd5442345g

**Comment by lsitu** [comment]
Hi Greg,
Yes. Currently we are matching for the mads scheme as well. So those non-fasting heading authority records won't be considered for a match. I'll update the matching rules to remove the matching of the mads scheme so that we can bring up those existing non-FAST heading authority records for linking.

**Comment by GregReser** [comment]
Great.  I'll test again when it's ready.

**Comment by lsitu** [comment]
@mcritchlow I've implemented the new rules for authority records lookup as suggested by @arwenhutt and @GregReser , see PR https://github.com/ucsdlib/damsmanager/pull/109. It's ready for review now and please merge it to the release branch if it looks good. Thanks.

**Comment by GregReser** [comment]
@lsitu can I test it in staging now?

**Comment by lsitu** [comment]
@GregReser No. We need to deploy it to staging once the PR is merged. I'll let you know when it's ready.

**Comment by GregReser** [comment]
OK, thanks

**Comment by lsitu** [comment]
@GregReser I've deployed damsmanager with the latest update for authority records lookup to staging. It's ready for test now. Thanks.

**Comment by GregReser** [comment]
@lsitu Good news: the matching logic is working correctly. 
Bad news: complex subjects are being created.  

The goal with this ticket is to bring the ingest process inline with our new approach to subjects for the new data model:

1. Eliminate pre-coordinated LCSH subjects
2. Eliminate dams:complexSubject.  As part of the subject conversion, we are splitting these into single subjects reconciled to a preferred vocabulary (FAST for topics, other types to be determined)
3. Use IDs, when available, for positive match (currently focusing on FAST in MARC, but other IDs will be used globally later)

From the log of my last test, it looks like new complexSubjects are being created.  Can you confirm these were new and not matched to existing ARKS?
[damslog-815936834.txt](https://github.com/ucsdlib/damsmanager/files/572544/damslog-815936834.txt)

ROGER RECORD: http://roger.ucsd.edu/record=b4163419~S9
651  7 California|zSan Diego.|2fast|0(OCoLC)fst01205232 
651  7 California|zSan Diego County.|2fast|0(OCoLC)fst01204290 

INGESTED OBJECT:  https://librarytest.ucsd.edu/dc/object/bd5032881t

NEW ARKS CREATED:
ComplexSubject: bd24731325: "California--San Diego County"
ComplexSubject: bd37359391: "California--San Diego"
Geographic: bd9879340j: "San Diego County"

Instead of dams:complexSubject, the two 651 values should result in two dams:geographic records.
"California--San Diego County"
"California--San Diego"

Most FAST headings are single terms, but some have multiple terms.  These are treated as a simple subject.  Any subfield between terms in a multi-part MARC FAST heading should be replaced with "--".

**Comment by lsitu** [comment]
@GregReser Yes, these fast headings subjects have multiple sub-fields which trigger our tool to consider it to be complex subjects. Is it possible that these multiple sub-fields are not in the same element type, for example one is Topic and the other is Geographic? How do we deal with it in this case?

**Comment by GregReser** [comment]
@lsitu That should not be a problem.  For MARC, any 6xx with "|2fast" should be an authorized FAST term, and will be OK.

For other ingest streams, we will develop a method to reconcile to a vocabulary before or during ingest (to be determined).  If we miss something, we still want simple subjects.  We will develop a method to review new DAMS authority records for class mismatch and other mistakes.  Subject review and maintenance will be an ongoing process until we retire.

**Comment by lsitu** [comment]
Hi Greg,
I've updated to convert those FAST Headings with multiple fields to simple subjects with the name of the first field element as the subject element name. Could you test it on staging again?

**Comment by GregReser** [comment]
@lsitu Geographics look great.  There should be two formGenre headings, which I don't see.
http://librarytest.ucsd.edu/dc/object/bd0118162r

ROGER RECORD: http://roger.ucsd.edu/record=b4163419~S9
655  7 Maps.|2fast|0(OCoLC)fst01423704 
655  7 Aerial photographs.|2fast|0(OCoLC)fst01423684

**Comment by lsitu** [comment]
@GregReser I think the issue is that we are trying to eliminate duplicate GenreForm elements in mods:genre and mods:subject/mods:genre. So when two such mods:genre element are detected , the mods:genre field (Maps and Aerial photographs in this case) are ignored.
I've fixed the logic to compare the siblings count as well. Could you try it now to see whether it works for you?

**Comment by GregReser** [comment]
@lsitu It's getting closer.  All testes worked except one.

dams:topic "Panama-California Exposition" not present.
DAMS INGEST: http://librarytest.ucsd.edu/dc/object/bd28485616
ROGER: http://roger.ucsd.edu/record=b6302708~S9
MISSING MARC: 611 27 Panama-California Exposition.|2fast|0(OCoLC)fst01407969

Is this because we didn't map 611?  That is an meeting name or event.

**Comment by lsitu** [comment]
@GregReser It looks like the metadata for MARC 611 is extracted as mads:Name, which should be mads:ConferenceName: http://lib-hydratail-staging.ucsd.edu:8080/dams/api/objects/bd28485616/export?es=no

I have the following names types for the MARC/MODS import tool to convert now: mods:name[@type='personal'], mods:name[@type='corporate'], and mods:name[@type='conference']. Is there any name types still missing, like family name?

**Comment by GregReser** [comment]
@lsitu We won't be using ConferenceName, so map MARC 611 to mods:name[@type='corporate'] and dams:corporateName.  We are taking this into consideration in our subject conversion project.

Family - there is no 6xx for family (thankfully, one less thing to map)

Be sure to remove trailing punctuation on the 611:
<mads:authoritativeLabel>Panama-California Exposition.</mads:authoritativeLabel>

**Comment by lsitu** [comment]
@GregReser I've updated JollyRoger to map field 611 to mods:name[@type='corporate'] and it'll be converted to mads:CorporateName in DAMS4. I also remove the tracing punctuation as you want. Could you test it on staging again? Thanks.

**Comment by GregReser** [comment]
@lsitu Excellent, it looks good.  Matching is right. Mapping is right.  No trailing punctuation.

DAMS INGEST: http://librarytest.ucsd.edu/dc/object/bd01522915
ROGER: http://roger.ucsd.edu/record=b9292834~S9

I think this is done.  Thanks for the good suggestions and for making all the changes.

**Comment by lsitu** [comment]
Thanks @GregReser. I am glad that it works.

**Comment by mcritchlow** [comment]
Great job @lsitu and thanks to @GregReser for all the QC and recommendations!

**Comment by GregReser** [comment]
@lsitu I think this is related to this issue, but not sure if it needs to be a new one or not.  Matt Peter's discovered that the CSV preview function is not including the 655 field.  All other FAST headings are working.

Roger: http://roger.ucsd.edu/record=b9292834~S9
655  7 History.|2fast|0(OCoLC)fst01411628 

The XML output contains the heading "History", but the CSV output does not.

[preview-1861538287.xlsx](https://github.com/ucsdlib/damsmanager/files/595939/preview-1861538287.xlsx)

**Comment by mcritchlow** [comment]
It would be nice to have this tracked as a separate ticket, but reference back to this one. @GregReser and @lsitu are you OK with that?

**Comment by lsitu** [comment]
@mcritchlow @GregReser It looks like something is missing from the CSV output and I am fine to work it out in another ticket and reference it back.
